### PR TITLE
allow data frame to change number of columns

### LINKED ISSUE #1198: Dataframe with headers throws error when new column is added
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If a `gr.Dataframe` is created with defined `headers`, then when a new column is created, it throws an error, since the backend tries to apply the original headers to the new dataframe which is of a different size

### Reproduction

Minimum example:

```py
gr.Interface(lambda x:x, 
             gr.Dataframe(headers=["a", "b", "c"]), 
             "dataframe").launch()
```

### Screenshot

_No response_

### Logs

_No response_

### System Info

```shell
gradio==2.9b25
```

### Severity

critical

### LINKED ISSUE #1430: Dynamic Dataset Size for Dataframe Component
is there any way we can dynamically handle consuming of the dataframe in the background without limiting users to upload a certain size? 

Like when we create the dataframe component we have to give a list to headers which should match the number of columns in a dataset in a CSV/TSV file you're dragging to UI. This is not good for below reasons:
- one thing in tabular workflows is that we have so many feature engineering going on, we add modify datasets by adding one hot encoded features or eliminate features according to shap or importance, it would be reaaaalllly essential to have such feature.
- not so many people know about the number of columns of their dataset in a CSV file beforehand.

maybe ping @pngwn  for this 🙂
thought: We can change the size through UI, so it should be fine for CSV/TSV file read to have.

### PR DESCRIPTION
**Pull request by pngwn**
Closes #1198. Closes #1430.

This PR changes the preprocess signature for Dataframe so we have access to the current headers rather than only the initial ones. I'm not sure if there is an impact outside of what I've tested, there may also be some test failures that I may not understand.

### CODE REVIEW

File: gradio/components.py
Comment (freddyaboulton) [inline gradio/components.py:2512]: I don't think it's correct to use the `self.col_count` defined when the component is created. If the headers are dynamic so is the `col_count`? 

In this repro, the function is only given access to the first column which is not the expected behavior imo

```python
import gradio as gr
import pandas as pd
import numpy as np

def return_empty(x):
    return np.atleast_1d(x.astype(float).mean(axis=0))


gr.Interface(return_empty,
             gr.Dataframe(type="numpy", headers=["a"]),
             gr.Dataframe(headers=["mean"])).launch()
```

File: gradio/components.py
Comment (freddyaboulton) [inline gradio/components.py:2500]: nit: Let's change the type hint here as well.

File: gradio/components.py
Comment (pngwn) [reply gradio/components.py:2512]: What was this logic ever even doing? This feels wrong regardless of the changes in this PR. Can we remove? @abidlabs @aliabid94

File: test/test_components.py
Comment (freddyaboulton) [inline test/test_components.py:1052]: Left some comments explaining why on slack but `self.assertEqual(output["Age"][1], 24)`  makes the diff easier to explain and justify. Not blocking though

File: gradio/components.py
Comment (freddyaboulton) [inline gradio/components.py:2538]: Awesome, make sense to delete `self.col_count`

File: test/test_components.py
Comment (freddyaboulton) [inline test/test_components.py:1146]: nit: Think we should remove this print statement prior to merge

File: demo/blocks_plug/run.py
Comment (freddyaboulton) [inline demo/blocks_plug/run.py:3]: nit: We're re-importing gradio here and we're not using pandas

File: gradio/components.py
Comment (freddyaboulton) [inline gradio/components.py:2392]: Nice

File: demo/blocks_plug/run.py
Comment (freddyaboulton) [reply demo/blocks_plug/run.py:3]: Actually, should we move this to another demo? I think `blocks_plug` is a nice demo for how to set named tabs and don't want to lose it unless there's a good reason for it.

**Review by freddyaboulton — APPROVED** [review]
Thanks for this feature @pngwn ! Python looks good to me. Left some minor nits (I think we should get rid of the print statements) but otherwise nothing blocking merge.

File: test/test_components.py
Comment (pngwn) [reply test/test_components.py:1146]: ops

File: demo/blocks_plug/run.py
Comment (pngwn) [reply demo/blocks_plug/run.py:3]: Yeah something weird happened after a merge and I'm not sure what i did when resolving the conflict. I'll check this against main to see what is what and create a new demo for the tab changing if necessary.

Comment (pngwn) [reply demo/blocks_plug/run.py]: reinstated, no idea what happened.

**Pull request merged**
